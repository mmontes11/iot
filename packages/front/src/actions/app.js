import {
  TOGGLE_HAMBURGER_MENU,
  TOGGLE_MAP_DIALOG,
  TOGGLE_LANGUAGE_SELECTOR,
  SET_SHOW_ERROR,
} from "constants/actionTypes/app";

export const toggleHamburgerMenu = () => dispatch => {
  dispatch({ type: TOGGLE_HAMBURGER_MENU });
};

export const toggleMapDialog = () => dispatch => {
  dispatch({ type: TOGGLE_MAP_DIALOG });
};

export const toggleLanguageSelector = () => dispatch => {
  dispatch({ type: TOGGLE_LANGUAGE_SELECTOR });
};

export const setShowError = showError => dispatch => {
  dispatch({ type: SET_SHOW_ERROR, showError });
};
