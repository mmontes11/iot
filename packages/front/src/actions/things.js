import {
  THINGS_UPDATED,
  THING_SELECTED,
  THINGS_REQUEST,
  THINGS_REQUEST_SUCCESS,
  THINGS_REQUEST_ERROR,
  SHOW_THING_NOT_FOUND_ERROR,
} from "constants/actionTypes/things";
import iotClient from "lib/iotClient";

export const getThings = () => dispatch => {
  dispatch({ type: THINGS_REQUEST });
  iotClient.thingsService
    .getThings()
    .then(res => {
      dispatch({ type: THINGS_REQUEST_SUCCESS, statusCode: res.statusCode, error: null });
      dispatch({ type: THINGS_UPDATED, things: res.body.things });
    })
    .catch(error => dispatch({ type: THINGS_REQUEST_ERROR, statusCode: error.statusCode, error }));
};

export const selectThing = thing => dispatch => {
  dispatch({ type: THING_SELECTED, thing });
};

export const showThingNotFoundError = showNotFoundError => dispatch => {
  dispatch({ type: SHOW_THING_NOT_FOUND_ERROR, showNotFoundError });
};
