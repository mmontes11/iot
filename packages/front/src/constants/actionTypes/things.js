export const THINGS_UPDATED = "THINGS_UPDATED";
export const THING_SELECTED = "THING_SELECTED";
export const THINGS_REQUEST = "THINGS_REQUEST";
export const THINGS_REQUEST_SUCCESS = "THINGS_REQUEST_SUCCESS";
export const THINGS_REQUEST_ERROR = "THINGS_REQUEST_ERROR";
export const SHOW_THING_NOT_FOUND_ERROR = "SHOW_THING_NOT_FOUND_ERROR";
