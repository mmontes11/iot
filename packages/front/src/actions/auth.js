import {
  IS_AUTH,
  SET_USERNAME,
  SET_PASSWORD,
  LOGIN_REQUEST,
  LOGIN_REQUEST_SUCCESS,
  LOGIN_REQUEST_ERROR,
} from "constants/actionTypes/auth";
import { setShowError } from "actions/app";
import iotClient from "lib/iotClient";

export const isAuth = () => dispatch => {
  iotClient.authService
    .isAuth()
    .then(isAuthVal => dispatch({ type: IS_AUTH, isAuth: isAuthVal }))
    .catch(() => dispatch({ type: IS_AUTH, isAuth: false }));
};

export const setUsername = username => dispatch => {
  dispatch({ type: SET_USERNAME, username });
};

export const setPassword = password => dispatch => {
  dispatch({ type: SET_PASSWORD, password });
};

export const login = () => (dispatch, getState) => {
  dispatch({ type: LOGIN_REQUEST });
  const { username, password } = getState().auth;
  iotClient.authService.setCredentials(username, password);
  iotClient.authService
    .getToken()
    .then(response => {
      dispatch({
        type: LOGIN_REQUEST_SUCCESS,
        statusCode: response.statusCode,
        error: null,
      });
      isAuth()(dispatch);
    })
    .catch(error => {
      dispatch({
        type: LOGIN_REQUEST_ERROR,
        statusCode: error.statusCode,
        error,
      });
      setShowError(true)(dispatch);
    });
};

export const logout = () => dispatch => {
  iotClient.authService
    .logout()
    .then(() => {
      isAuth()(dispatch);
    })
    .catch(() => {
      isAuth()(dispatch);
    });
};
