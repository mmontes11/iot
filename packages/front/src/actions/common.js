import { RESET } from "constants/actionTypes/common";

export const reset = preserveError => dispatch => {
  dispatch({ type: RESET, preserveError });
};
