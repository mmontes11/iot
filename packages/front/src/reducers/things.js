import { THINGS_UPDATED, THING_SELECTED, SHOW_THING_NOT_FOUND_ERROR } from "constants/actionTypes/things";
import { RESET } from "constants/actionTypes/common";

export const initialState = {
  items: [],
  selectedItem: null,
  showNotFoundError: false,
};

export default (state = initialState, { type, things, thing, showNotFoundError }) => {
  switch (type) {
    case THINGS_UPDATED:
      return { ...state, items: things };
    case THING_SELECTED:
      if (state.selectedItem !== null && state.selectedItem.name === thing.name) {
        return { ...state, selectedItem: null };
      }
      return { ...state, selectedItem: thing };
    case SHOW_THING_NOT_FOUND_ERROR:
      return { ...state, showNotFoundError };
    case RESET:
      return { ...initialState, items: state.items };
    default:
      return state;
  }
};
