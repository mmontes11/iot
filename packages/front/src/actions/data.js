import {
  DATA_UPDATED,
  DATA_REQUEST,
  DATA_REQUEST_SUCCESS,
  DATA_REQUEST_ERROR,
  ADD_DATA_ITEM,
  RESET_DATA,
} from "constants/actionTypes/data";
import { EVENT_TYPE, MEASUREMENT_TYPE } from "constants/observationTypes";
import iotClient from "lib/iotClient";
import { SocketController } from "helpers/socketController";
import { THING_FILTER_TYPE, DATE_FILTER_TYPE } from "constants/filterTypes";
import { RESET } from "constants/actionTypes/common";
import * as fromState from "reducers";
import { TYPE, OBSERVATION, GROUPBY, THING } from "constants/params";

const getParams = (filterItems, observation, groupBy, thing, timePeriod, startDate, endDate) => {
  let params = {
    type: observation,
  };
  if (groupBy) {
    params = {
      ...params,
      groupBy,
    };
  }
  if (filterItems.includes(THING_FILTER_TYPE) && thing) {
    params = {
      ...params,
      thing,
    };
  }
  if (filterItems.includes(DATE_FILTER_TYPE)) {
    if (timePeriod) {
      params = {
        ...params,
        timePeriod,
      };
    } else {
      if (startDate) {
        params = {
          ...params,
          startDate: startDate.toISOString(),
        };
      }
      if (endDate) {
        params = {
          ...params,
          endDate: endDate.toISOString(),
        };
      }
    }
  }
  return params;
};

const requestStats = (
  filterItems,
  type,
  observation,
  thing,
  timePeriod,
  startDate,
  endDate,
  onStart,
  onSuccess,
  onError,
) => {
  const params = getParams(filterItems, observation, null, thing, timePeriod, startDate, endDate);
  if (type === EVENT_TYPE) {
    onStart();
    iotClient.eventService
      .getStats(params)
      .then(res => onSuccess(res))
      .catch(err => onError(err));
  } else if (type === MEASUREMENT_TYPE) {
    onStart();
    iotClient.measurementService
      .getStats(params)
      .then(res => onSuccess(res))
      .catch(err => onError(err));
  } else {
    onError(new Error("Invalid type"));
  }
};

export const getStats = () => (dispatch, getState) => {
  const state = getState();
  const filterItems = state.filters.items;
  const type = fromState.getParam(state, TYPE).selectedItem;
  const observation = fromState.getParam(state, OBSERVATION).selectedItem;
  const thing = state.filters.thingFilter.selectedItem;
  const {
    timePeriod: { selectedItem: selectedTimePeriod },
    custom: { startDate, endDate },
  } = state.filters.dateFilter;
  if (!type || !observation) {
    return;
  }
  requestStats(
    filterItems,
    type,
    observation,
    thing,
    selectedTimePeriod,
    startDate,
    endDate,
    () => dispatch({ type: DATA_REQUEST }),
    res => {
      dispatch({ type: DATA_REQUEST_SUCCESS });
      dispatch({ type: DATA_UPDATED, items: res.body.stats });
    },
    () => {
      dispatch({ type: DATA_REQUEST_ERROR });
      dispatch({ type: RESET, preserveError: true });
    },
  );
};

const requestData = (
  filterItems,
  type,
  observation,
  groupBy,
  thing,
  timePeriod,
  startDate,
  endDate,
  onStart,
  onSuccess,
  onError,
) => {
  const params = getParams(filterItems, observation, groupBy, thing, timePeriod, startDate, endDate);
  if (type === EVENT_TYPE) {
    onStart();
    iotClient.eventService
      .getData(params)
      .then(res => onSuccess(res))
      .catch(err => onError(err));
  } else if (type === MEASUREMENT_TYPE) {
    onStart();
    iotClient.measurementService
      .getData(params)
      .then(res => onSuccess(res))
      .catch(err => onError(err));
  } else {
    onError(new Error("Invalid type"));
  }
};

export const getData = () => (dispatch, getState) => {
  const state = getState();
  const filterItems = state.filters.items;
  const type = fromState.getParam(state, TYPE).selectedItem;
  const observation = fromState.getParam(state, OBSERVATION).selectedItem;
  const groupBy = fromState.getParam(state, GROUPBY).selectedItem;
  const thing = state.filters.thingFilter.selectedItem;
  const {
    timePeriod: { selectedItem: selectedTimePeriod },
    custom: { startDate, endDate },
  } = state.filters.dateFilter;
  if (!type || !observation) {
    return;
  }
  requestData(
    filterItems,
    type,
    observation,
    groupBy,
    thing,
    selectedTimePeriod,
    startDate,
    endDate,
    () => dispatch({ type: DATA_REQUEST }),
    ({ body: { data, things } }) => {
      dispatch({ type: DATA_REQUEST_SUCCESS });
      dispatch({ type: DATA_UPDATED, items: data, things });
    },
    () => {
      dispatch({ type: DATA_REQUEST_ERROR });
      dispatch({ type: RESET, preserveError: true });
    },
  );
};

let socketController = null;

export const finishRealTimeData = () => () => {
  if (socketController !== null) {
    socketController.close();
  }
};

export const startRealTimeData = () => (dispatch, getState) => {
  const state = getState();
  const thing = fromState.getParam(state, THING).selectedItem;
  const type = fromState.getParam(state, TYPE).selectedItem;
  if (!thing || !type) {
    return;
  }
  finishRealTimeData()();
  const onData = data => {
    const {
      data: { items },
    } = getState();
    if (items.length === 0) {
      dispatch({ type: DATA_REQUEST_SUCCESS });
    }
    dispatch({ type: ADD_DATA_ITEM, item: data });
  };
  const onError = error => {
    dispatch({ type: DATA_REQUEST_ERROR, error });
    dispatch({ type: RESET, preserveError: true });
  };
  socketController = new SocketController(thing, type, onData, onError);
  socketController.listen();
  dispatch({ type: DATA_REQUEST });
};

export const resetData = () => dispatch => {
  dispatch({ type: RESET_DATA });
};
