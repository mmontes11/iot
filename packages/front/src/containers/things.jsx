import React from "react";
import { connect } from "react-redux";
import { withRouter } from "react-router-dom";
import { compose } from "recompose";
import PropTypes from "prop-types";
import ThingItem from "components/thingItem";
import ThingDetail from "components/thingDetail";
import Loader from "components/loader";
import Modal from "components/modal";
import * as thingActions from "actions/things";
import * as fromState from "reducers";
import { withResetOnUnmount } from "hocs/resetOnUnmount";
import { injectIntl, intlShape } from "react-intl";
import { EVENT_TYPE, MEASUREMENT_TYPE } from "constants/observationTypes";
import { defaultGroupBy } from "config/params";

class Things extends React.Component {
  componentDidMount() {
    if (this.props.things.length === 0) {
      this.props.getThings();
    }
  }
  componentDidUpdate() {
    const {
      selectedThing,
      things,
      match: {
        params: { thing: thingName },
      },
    } = this.props;
    if (selectedThing === null && things.length > 0 && thingName) {
      const thing = things.find(t => t.name === thingName);
      if (thing) {
        this._selectThing(thing);
      } else {
        this._resetPathToRoot();
        this.props.showNotFoundError(true);
      }
    }
  }
  _isSelected = thing => this.props.selectedThing !== null && this.props.selectedThing.name === thing.name;
  _selectThing = thing => {
    this.props.selectThing(thing);
    if (this._isSelected(thing)) {
      this._resetPathToRoot();
    } else {
      this.props.history.push(`/things/${thing.name}`);
    }
  };
  _resetPathToRoot = () => {
    this.props.history.push("/things");
  };
  _onStatsClick = (type, thing) => {
    const supportedEvents = thing.supportedObservationTypes.event;
    const supportedMeasurements = thing.supportedObservationTypes.measurement;
    let url;
    if (type === EVENT_TYPE && supportedEvents && supportedEvents.length > 0) {
      url = `/stats/${type}/${supportedEvents[0]}?thing=${thing.name}&timePeriod=day`;
    } else if (type === MEASUREMENT_TYPE && supportedMeasurements && supportedMeasurements.length > 0) {
      url = `/stats/${type}/${supportedMeasurements[0]}?thing=${thing.name}&timePeriod=day`;
    }
    if (url) {
      this.props.history.push(url);
    }
  };
  _onDataClick = (type, observation, thing) => {
    this.props.history.push(`/data/${type}/${observation}?thing=${thing.name}&groupBy=${defaultGroupBy}`);
  };
  render() {
    const {
      intl: { formatMessage },
      isLoading,
      things,
      selectedThing,
      shouldShowNotFoundError,
      showNotFoundError,
    } = this.props;
    if (isLoading && things.length === 0) {
      return <Loader />;
    }
    return (
      <div>
        <div className="container is-fluid section">
          <div className="columns">
            <div className="column is-one-quarter">
              {things.map(thing => (
                <ThingItem
                  key={thing.name}
                  name={thing.name}
                  isSelected={this._isSelected(thing)}
                  onClick={() => this._selectThing(thing)}
                />
              ))}
            </div>
            {selectedThing && (
              <div className="column is-three-quarters">
                <ThingDetail
                  thing={selectedThing}
                  onStatsClick={type => this._onStatsClick(type, selectedThing)}
                  onDataClick={(type, observation) => this._onDataClick(type, observation, selectedThing)}
                />
              </div>
            )}
          </div>
        </div>
        <Modal
          isActive={shouldShowNotFoundError}
          onCloseClick={() => showNotFoundError(false)}
          messageStyle="is-danger"
          title={formatMessage({ id: "Error" })}
          subTitle={formatMessage({ id: "Thing not found" })}
        />
      </div>
    );
  }
}

Things.propTypes = {
  intl: intlShape.isRequired,
  getThings: PropTypes.func.isRequired,
  selectThing: PropTypes.func.isRequired,
  selectedThing: PropTypes.shape({
    name: PropTypes.string.isRequired,
  }),
  isLoading: PropTypes.bool.isRequired,
  things: PropTypes.arrayOf(PropTypes.shape({})),
  shouldShowNotFoundError: PropTypes.bool,
  showNotFoundError: PropTypes.func.isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }).isRequired,
  match: PropTypes.shape({}).isRequired,
};

Things.defaultProps = {
  selectedThing: null,
  things: [],
  shouldShowNotFoundError: false,
};

const withConnect = connect(
  state => ({
    isLoading: fromState.isLoading(state),
    things: state.things.items,
    selectedThing: state.things.selectedItem,
    shouldShowNotFoundError: state.things.showNotFoundError,
  }),
  {
    getThings: thingActions.getThings,
    selectThing: thingActions.selectThing,
    showNotFoundError: thingActions.showThingNotFoundError,
  },
);

export default compose(
  withConnect,
  withRouter,
  withResetOnUnmount,
  injectIntl,
)(Things);
