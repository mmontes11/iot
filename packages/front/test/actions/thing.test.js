import { getThings, selectThing } from "actions/things";

describe("actions/auth", () => {
  const testThunk = thunk => {
    const dispatch = jest.fn();
    thunk(dispatch);
    expect(thunk).toBeAThunk();
    expect(thunk).toMatchSnapshot();
  };
  it("dispatches a getThings action", () => {
    testThunk(getThings());
  });
  it("dispatches a selectThing action", () => {
    testThunk(selectThing());
  });
});
