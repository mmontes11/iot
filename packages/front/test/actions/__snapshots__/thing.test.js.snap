// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions/auth dispatches a getThings action 1`] = `[Function]`;

exports[`actions/auth dispatches a selectThing action 1`] = `[Function]`;
